opencv/opencv
# G-API: Introduce ONNX backend for Inference

- Basic operations are implemented (Infer, -ROI, -List, -List2);
- Implemented automatic preprocessing for ONNX models;
- Test suite is extended with `OPENCV_GAPI_ONNX_MODEL_PATH` env for test data
  (test data is an ONNX Model Zoo repo snapshot);
- Fixed kernel lookup logic in core G-API:
  - Lookup NN kernels not in the default package, but in the associated
    backend's aux package. Now two NN backends can work in the same graph.
- Added Infer SSD demo and a combined ONNX/IE demo;

<cut/>

### How to build:

Build & install the ONNX RT (currently tested with v1.5.1):
```bash
$ git clone --recursive https://github.com/microsoft/onnxruntime.git
$ cd onnxruntime
$ git checkout v1.5.1
$ git submodule update --init
$ ./build.sh --config Release --build_shared_lib --parallel \
$     --cmake_extra_defines CMAKE_INSTALL_PREFIX=install
$ cd build/Linux/Release
$ make install
```

...then specify extra options to OpenCV CMake:
```bash
$ cmake /path/to/opencv -DWITH_ONNX=ON -DORT_INSTALL_DIR=path-to-onnxrt-install
```

### How to test:

Clone ONNX Model Zoo & selected models
```bash
git clone --recursive https://github.com/onnx/models.git
cd models
git lfs pull --include=vision/object_detection_segmentation/ssd-mobilenetv1/model/ssd_mobilenet_v1_10.onnx --exclude=""
git lfs pull --include=vision/classification/squeezenet/model/squeezenet1.0-9.onnx --exclude=""
git lfs pull --include=vision/body_analysis/emotion_ferplus/model/emotion-ferplus-8.onnx --exclude=""
```
When running G-API tests, specify `OPENCV_GAPI_ONNX_MODEL_PATH` environment variable to the model repo root. Expected result:
```
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from ONNX
[ RUN      ] ONNX.Infer
[       OK ] ONNX.Infer (37 ms)
[ RUN      ] ONNX.InferROI
[       OK ] ONNX.InferROI (26 ms)
[ RUN      ] ONNX.InferROIList
[       OK ] ONNX.InferROIList (44 ms)
[ RUN      ] ONNX.Infer2ROIList
[       OK ] ONNX.Infer2ROIList (43 ms)
[----------] 4 tests from ONNX (150 ms total)
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-1
build_image:Custom=ubuntu-onnx:20.04
test_modules:Custom=gapi
test_filter:Custom=*ONNX*
```
